cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,282 @@
+    else if (((dash_count > 0 && dash_count <= 2 && slash_count == 0) ||
+              (dash_count == 0 && slash_count > 0 && slash_count <= 2)) &&
+             colon_count <= 2) {
+      atomicAdd(&column_info->datetime_count, 1);
+    }
+  }  // grid-stride for loop
+}
+
+/**
+ * @brief Constructs column type histogram for a given column string input `data`.
+ *
+ * @tparam ColumnStringIter Iterator type whose `value_type` is a
+ * `thrust::tuple<offset_t, length_t>`, where `offset_t` and `length_t` are of integral type and
+ * `offset_t` needs to be convertible to `std::size_t`.
+ *
+ * @param options View of inference options
+ * @param data JSON string input
+ * @param column_strings_begin The begining of an offset-length tuple sequence
+ * @param size Size of the string input
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A histogram containing column-specific type counters
+ */
+template <typename ColumnStringIter>
+cudf::io::column_type_histogram infer_column_type(json_parse_options_view const& options,
+                                                  cudf::device_span<char const> data,
+                                                  ColumnStringIter column_strings_begin,
+                                                  std::size_t const size,
+                                                  rmm::cuda_stream_view stream)
+{
+  constexpr int block_size = 128;
+
+  auto const grid_size = (size + block_size - 1) / block_size;
+  auto d_column_info   = rmm::device_scalar<cudf::io::column_type_histogram>(stream);
+  CUDF_CUDA_TRY(cudaMemsetAsync(
+    d_column_info.data(), 0, sizeof(cudf::io::column_type_histogram), stream.value()));
+
+  infer_column_type_kernel<<<grid_size, block_size, 0, stream.value()>>>(
+    options, data, column_strings_begin, size, d_column_info.data());
+
+  return d_column_info.value(stream);
+}
+
+/**
+ * @brief Infers data type for a given JSON string input `data`.
+ *
+ * @throw cudf::logic_error if input size is 0
+ * @throw cudf::logic_error if data type inference failed
+ *
+ * @tparam ColumnStringIter Iterator type whose `value_type` is convertible to
+ * `thrust::tuple<device_span, string_view>`
+ *
+ * @param options View of inference options
+ * @param data JSON string input
+ * @param column_strings_begin The begining of an offset-length tuple sequence
+ * @param omission_null_count Number of omitted nulls
+ * @param size Size of the string input
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return The inferred data type
+ */
+template <typename ColumnStringIter>
+cudf::data_type infer_data_type(json_parse_options_view const& options,
+                                device_span<char const> data,
+                                ColumnStringIter column_strings_begin,
+                                cudf::size_type omission_null_count,
+                                std::size_t const size,
```

Comment:
I don't get how these parameters work together. If there are omitted elements, they will not be present in the input strings. This should mean that size does not include such elements, which in turn means that we don't need `omission_null_count` to determine "all null" case.

Reply:
@vuule is right. Sorry, I screwed up here. We can revert the changes of the ommitted null count.
Omissions will be passed as zero-length strings here and I had forgotton to make the empty string part of the trie_na.

Reply:
Reverted